### ISSUE TITLE
node stops syncing, no errors

### ISSUE THREAD

**Issue by j4ys0n** [CityOfZion/neo-python#492]
### Current behavior

- What is the problem? Node stops syncing new blocks. If I stop/start the node, it continues syncing, but it will sit at the same block for days if that's not done.
- Include a screenshot if it makes sense.

```
[I 180704 00:36:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:36:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:36:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:37:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:37:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:37:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:37:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:38:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:38:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:38:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:38:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:39:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:39:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:39:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:39:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:40:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:40:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:40:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:40:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:41:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:41:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:41:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:41:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:42:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:42:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:42:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:42:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:43:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:43:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:43:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:43:46 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:44:01 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:44:16 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:44:31 threading:864] [MainNet] Block 2278690 / 2464138
[I 180704 00:44:46 threading:864] [MainNet] Block 2278690 / 2464138
```

### Expected behavior

- What should be happening? Node should sync new blocks
- Include a screenshot if it makes sense.

### How to reproduce

- Please explain the steps to reproduce the problem. If you are having an issue with your machine or build tools, the issue belongs on another repository as that is outside of the scope of this project.

### Your environment

Let us know in what environment you're running into the issue:

- OS: (Windows version/Linux distro/OSX version) 
- neo-python version: (hint: use `np-prompt --version`)
- Python version: (hint: `python -V`)

latest build from project's docker file.

**Comment by metachris** [comment]
Thanks for the report.

I've also experienced this a few days ago when syncing the chain on 2 servers. One just kept going, the other got stuck at some point. Looking into it.

**Comment by j4ys0n** [comment]
We don't need gas to run a node do we? Seeing this message sometimes:
`[E 180705 23:11:43 ApplicationEngine:266] NOT ENOUGH GAS`

**Comment by metachris** [comment]
No you don't need GAS to run a node. This is error messages of other people's smart contract invocations.

**Comment by j4ys0n** [comment]
That’s what I thought, just trying to figure out why it stops syncing. I’ve got another node I started to sync, but still quite a bit of time left before the sync is complete.

**Comment by dethos** [comment]
I also can confirm the issue. I have 2 nodes (one on `test net` the the `main net` both of them get stuck. 

Restarting the node helps, but it stops syncing after a few moments.

**Comment by dethos** [comment]
Just to let you know, after the last 0.7.5 update things improved greatly. In ~4 days since the update, the issue just occurred once on my nodes.

**Comment by ixje** [comment]
It should become even better once this is merged: https://github.com/CityOfZion/neo-python/pull/520
Maybe we should wait until the above PR is in master and test a couple more days with that included

**Comment by dethos** [comment]
After a few more days, I've the nodes got stuck again (a few times, but not as before) so I'm not sure this is completely solved yet. Will wait for the #520 and test again.

**Comment by ixje** [comment]
@dethos what branch were you on when this happened? #520 was merged the same day, so it will be interesting to learn if you testing with or without those enhancements

**Comment by dethos** [comment]
@ixje without those enhancements. My production nodes are running `0.7.5` (I try to follow the official releases).

**Comment by metachris** [comment]
Running the development branch is what I prefer and would suggest :)

**Comment by ixje** [comment]
Has anyone tried running `neo-python` on a branch with #520 in it? If so please share your results so we can decide whether to keep the issue open or close it.

**Comment by tomascarvalho** [comment]
@ixje I'm running `neo-python` on the `testnet` in the `development` branch, which, if I'm not wrong, is already patched with #520 and it is still getting stuck.

**Comment by localhuman** [comment]
@tomascarvalho this is a different issue, which should be solved by #540

**Comment by j4ys0n** [comment]
I'm still seeing this issue.

Running 0.7.6 and 0.8.0.

`[I 181004 16:44:59 threading:864] [MainNet] Block 2612462 / 2805804`

We're running 4 nodes and they're all stuck between blocks 2.6 milllion and 2.78 million. Restarting the nodes does not cause them to start syncing again.

There is an error on one of the nodes, I'm not sure if it's related.

```
[I 181003 18:21:56 threading:864] [MainNet] Block 2786143 / 2804116
[I 181003 18:22:08 defer:150] Could not persist block argument out of range 
Unhandled error in Deferred:
[C 181003 18:22:08 _observer:131] Unhandled error in Deferred:

[C 181003 18:22:08 _observer:131]
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 150, in maybeDeferred
        result = f(*args, **kw)
      File "/neo-python/neo/Implementations/Blockchains/LevelDB/LevelDBBlockchain.py", line 852, in PersistBlocks
        raise e
      File "/neo-python/neo/Implementations/Blockchains/LevelDB/LevelDBBlockchain.py", line 848, in PersistBlocks
        self.OnPersistCompleted(block)
      File "/neo-python/neo/Core/Blockchain.py", line 459, in OnPersistCompleted
        self.PersistCompleted.on_change(block)
      File "/usr/local/lib/python3.6/dist-packages/events/events.py", line 95, in __call__
        f(*a, **kw)
      File "/neo-python/neo/Implementations/Notifications/LevelDB/NotificationDB.py", line 143, in on_persist_completed
        hash_data = evt.ToByteArray()
      File "/neo-python/neo/SmartContract/SmartContractEvent.py", line 120, in ToByteArray
        self.Serialize(writer)
      File "/neo-python/neo/SmartContract/SmartContractEvent.py", line 85, in Serialize
        self.SerializePayload(writer)
      File "/neo-python/neo/SmartContract/SmartContractEvent.py", line 254, in SerializePayload
        writer.WriteVarInt(self.amount)
      File "/usr/local/lib/python3.6/dist-packages/neocore/IO/BinaryWriter.py", line 339, in WriteVarInt
        return self.WriteUInt64(value, endian)
      File "/usr/local/lib/python3.6/dist-packages/neocore/IO/BinaryWriter.py", line 272, in WriteUInt64
        return self.pack('%sQ' % endian, value)
      File "/usr/local/lib/python3.6/dist-packages/neocore/IO/BinaryWriter.py", line 118, in pack
        return self.WriteBytes(struct.pack(fmt, data), unhex=False)
    struct.error: argument out of range
[I 181003 18:22:11 threading:864] [MainNet] Block 2786173 / 2804116
[I 181003 18:22:13 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7efbdc147fd0>
[I 181003 18:22:14 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7efbfbb3f898>
[I 181003 18:22:15 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7efbdc149240>
[I 181003 18:22:26 threading:864] [MainNet] Block 2786173 / 2804117
```

**Comment by j4ys0n** [comment]
Any update on this?

**Comment by ixje** [comment]
Are they all stuck on the same block? If so, which block?

**Comment by j4ys0n** [comment]
No, different blocks. All info is in the post above: https://github.com/CityOfZion/neo-python/issues/492#issuecomment-427092328

**Comment by ixje** [comment]
It's hard to say what goes on there, I'm syncing up now from `2784525`, will try to see what happens. 
PS: if you updated to `0.8.0` you'll have to resync your chain (or use `np-bootstrap -m`) because there was an important fix to storage that could caused inconsistencies in stored balances.

**Comment by ixje** [comment]
@j4ys0n for what it's worth, I just synced from `2784525` up to `2791504` and progressing without errors, just 1 warning being;
> [W 181010 18:31:22 SmartContractEvent:276] Writing Payload value amount greater than ulong long is not allowed.  Setting to ulong long max

That's it.

**Comment by ixje** [comment]
Ok I've encountered a scenario that could cause 'hanging'. 

It can happen that we request blocks from our connected nodes, the nodes we request them from are alive (i.e. the connection is established), but they don't give us the requested blocks. Then neo-python just hangs there because it doesn't a timeout mechanism for answering to our request. If we would have that and track some `health` like stats we could go find a new node that does answer to our requests.

**Comment by j4ys0n** [comment]
that sounds like it could be the issue :)

is there a way to manually add new peer connections?

**Comment by ixje** [comment]
You could add them to the seedlist:
https://github.com/CityOfZion/neo-python/blob/1790581bfb9c91e92814fe6624997f90c08f989f/neo/data/protocol.mainnet.json#L14-L30

Another thing to keep in mind is that by default `neo-python` only accepts 5 peers. So if the first 5 in the seedlist are alive but not delivering the required data, then you'd still be stuck. You can increase this by adding the `--maxpeers` option on the command line.

**Comment by j4ys0n** [comment]
That doesn't seem to have done it. I built a new docker image with the latest version (which includes that seed list you posted) and set `--maxpeers 50` the blockchain stopped syncing at ~22%. I just restarted the container and it's syncing again, but I don't think the seed list and upping the peers fixed the issue.

**Comment by j4ys0n** [comment]
```
[I 181025 16:49:53 threading:864] [MainNet] Block 2786175 / 2891433
[I 181025 16:49:55 defer:150] Could not persist block argument out of range
Unhandled error in Deferred:
[C 181025 16:49:55 _observer:131] Unhandled error in Deferred:

[C 181025 16:49:55 _observer:131]
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 150, in maybeDeferred
        result = f(*args, **kw)
      File "/neo-python/neo/Implementations/Blockchains/LevelDB/LevelDBBlockchain.py", line 852, in PersistBlocks
        raise e
      File "/neo-python/neo/Implementations/Blockchains/LevelDB/LevelDBBlockchain.py", line 848, in PersistBlocks
        self.OnPersistCompleted(block)
      File "/neo-python/neo/Core/Blockchain.py", line 459, in OnPersistCompleted
        self.PersistCompleted.on_change(block)
      File "/usr/local/lib/python3.6/dist-packages/events/events.py", line 95, in __call__
        f(*a, **kw)
      File "/neo-python/neo/Implementations/Notifications/LevelDB/NotificationDB.py", line 143, in on_persist_completed
        hash_data = evt.ToByteArray()
      File "/neo-python/neo/SmartContract/SmartContractEvent.py", line 120, in ToByteArray
        self.Serialize(writer)
      File "/neo-python/neo/SmartContract/SmartContractEvent.py", line 85, in Serialize
        self.SerializePayload(writer)
      File "/neo-python/neo/SmartContract/SmartContractEvent.py", line 254, in SerializePayload
        writer.WriteVarInt(self.amount)
      File "/usr/local/lib/python3.6/dist-packages/neocore/IO/BinaryWriter.py", line 339, in WriteVarInt
        return self.WriteUInt64(value, endian)
      File "/usr/local/lib/python3.6/dist-packages/neocore/IO/BinaryWriter.py", line 272, in WriteUInt64
        return self.pack('%sQ' % endian, value)
      File "/usr/local/lib/python3.6/dist-packages/neocore/IO/BinaryWriter.py", line 118, in pack
        return self.WriteBytes(struct.pack(fmt, data), unhex=False)
    struct.error: argument out of range
```

**Comment by ixje** [comment]
@j4ys0n can you please let us know which version that error occurred on? I know there has been some  digging and attempts to resolve all "Unhandled error in Deferred" errors. I would like to check if you're using a release pre of post those fixes. thanks

**Comment by j4ys0n** [comment]
yep, just checked it - `NEO-PYTHON:0.7.1`

**Comment by j4ys0n** [comment]
i just copied out the data dir from that node, built a new docker image with 0.8.1 and mounted the existing chain on the new node and it appears to be syncing again. it was stuck around block 2786210. 0.7.1 would sync a few blocks then give that Deferred error above. This appears to be working now with 0.8.1... fingers crossed.

The only issue is syncing from scratch, I started 2 nodes with 0.8.1 to sync from scratch and they're only at block ~600k, after syncing for 9 days. They're still going, but very slowly.

**Comment by ThomasLobker** [comment]
I'm also trying to do a full sync from scratch with `neo-python` version `0.8.1` (with two completely separate nodes) since about a week. It's been a nightmare to be honest. It's now at about 1.3M blocks (so even before any NEP-5 tokens) and progress is extremely slow. When I'm starting the day I'm checking progress and sometimes it's stuck at the same block for hours, when I restart the node it seems to continue again for a while, but never for long.

**Comment by ixje** [comment]
@j4ys0n I'm not sure if that will work without unexpected results due to a fix to the way data is stored. I'm sure it will continue syncing, but it is likely to have wrong data stored now.

@ThomasLobker yeah I've been reading reports of this more and more lately, and sometimes experiencing this myself (even on a privatenet). In about 2 weeks I'm going to have a lot of time at hand and definitely want to dive into this. I don't expect it to be easy, but it is not performing well atm and we need to fix that a.s.a.p.

**Comment by ThomasLobker** [comment]
I forgot to mention that the node was also crashing a couple of times, and giving a lot of warnings and error messages. Tomorrow back at the office I can send them over.

Right now I'm trying a fresh full sync from scratch again, with only one peer (neo-cli 2.9.0) on localhost. Let's see if this works better.

**Comment by ixje** [comment]
great, would love to receive those logs

**Comment by xxedocxx** [comment]
Progress: 2860777 / 2893200
Block-cache length 22
Blocks since program start 1165
Time elapsed 358.3456694833333 mins
Blocks per min 3.2510508685083583
TPS: 0.8365758508134455

I have received 10000 blocks last 2 days.

**Comment by ThomasLobker** [comment]
I managed to improve the situation a lot, but using only **one peer** (`neo-cli 2.9.0`) that was already fully synced and that runs on the same machine, over loopback. I was so far able to do a full sync up to about 1.7M blocks in the last 18 hours or so. The issue is not my local internet connection. The syncing progress did halt a couple of times, but simply restarting the node fixed that.

> Tomorrow back at the office I can send them over.

Unfortunately I can't get this error message back. If I recall correctly, it said something about reaching maximum levels in recursion, and then it just crashed.

**Comment by ixje** [comment]
I just watched syncing mainnet from scratch a bit and there's just a lot of nodes that don't respond to neo-python's request, so we keep repeating the request to no end. I have a stale count of 26000 (thousand) as we speak. One of the things I want to add is new node discovery so we can discard nodes that don't respond to our block requests (likely because they're overloaded) after x number of times.

**Comment by xxedocxx** [comment]
I just used np-bootstrap with the latest merged. It finally works!

**Comment by ThomasLobker** [comment]
I noticed that it's always resetting blocks before it starts to hang.

When `self.leader.ResetBlockRequestsAndCache()` is called [here](https://github.com/CityOfZion/neo-python/blob/6bda86257460078e08543217b0d8905c16be29ab/neo/Network/NeoNode.py#L308) it's clearing the cache, specifically `BC.Default()._block_cache`. It happens once in a while that this happens right between receiving the block from the remote peer and [trying to persist the block here](https://github.com/CityOfZion/neo-python/blob/6bda86257460078e08543217b0d8905c16be29ab/neo/Implementations/Blockchains/LevelDB/LevelDBBlockchain.py#L818). So basically the missing block is received and removed right before processing. For some reason that block is not requested or received again, so `neo-python` is waiting infinitely for that block.

**Workaround:**

```diff
diff --git a/neo/Network/NodeLeader.py b/neo/Network/NodeLeader.py
index 4a77c24..49c6b5c 100644
--- a/neo/Network/NodeLeader.py
+++ b/neo/Network/NodeLeader.py
@@ -238,7 +238,6 @@ class NodeLeader:
         for p in self.Peers:
             p.myblockrequests = set()
         BC.Default().ResetBlockRequests()
-        BC.Default()._block_cache = {}
 
     def InventoryReceived(self, inventory):
         """
```

**Comment by j4ys0n** [comment]
I've got 2 nodes that i've been trying to sync for about 2 weeks now (0.8.1) and they're stuck again. they've only made it to block 101692 and 766257 respectively.

**Comment by j4ys0n** [comment]
I've had 2 nodes stuck for a few days now. It's been 2+ weeks and they're still not fully synced. Both are 0.8.1. I see a lot of messages like this: `Could not find transaction for hash b'81871e3cd4b9e1bd2cecf863ef9884d934f8d851dc57acf5b628a4fac1b5f128'` I'm going to update them to 0.8.2, saw that was just released.

**Comment by ThomasLobker** [comment]
@j4ys0n make sure you use the development branch. Also please notice that my proposed workaround is not yet merged, so you should manually remove [line 243 from NodeLeader.py](https://github.com/CityOfZion/neo-python/blob/development/neo/Network/NodeLeader.py#L243). Thanks for testing!

**Comment by ixje** [comment]
should be resolved in `0.9.0`. Please open an new issue if it does not.
